File: sdks/python/src/verinode/services/verification.py
"""
Verification service for managing proof verifications.
"""

import logging
from typing import Optional, List, Dict, Any
from ..types import (
    Verification, VerificationCreateRequest, VerificationStatus,
    QueryOptions, PaginatedResponse
)
from ..exceptions import VerinodeError, VerinodeAPIError
from ..utils import HTTPClient


class VerificationService:
    """
    Service for managing verifications in the Verinode system.
    
    Provides methods to create, read, update, and query verifications.
    """
    
    def __init__(self, http_client: HTTPClient, config):
        """
        Initialize verification service.
        
        Args:
            http_client: HTTP client for API requests
            config: Configuration object
        """
        self.http_client = http_client
        self.config = config
        self.logger = logging.getLogger(__name__)
    
    async def create(self, request: VerificationCreateRequest) -> Verification:
        """
        Create a new verification.
        
        Args:
            request: Verification creation request
            
        Returns:
            Created verification
            
        Raises:
            VerinodeError: If verification creation fails
        """
        try:
            response = await self.http_client.post(
                "/verifications",
                json=request.dict(exclude_none=True)
            )
            
            if response.get("success"):
                verification_data = response["data"]
                verification = Verification(**verification_data)
                self.logger.info(f"Created verification: {verification.id}")
                return verification
            else:
                raise VerinodeError(
                    response.get("error", "Failed to create verification"),
                    error_code="VERIFICATION_CREATE_FAILED"
                )
                
        except Exception as e:
            if isinstance(e, VerinodeError):
                raise
            raise VerinodeError(f"Failed to create verification: {str(e)}")
    
    async def get(self, verification_id: str) -> Verification:
        """
        Get a verification by ID.
        
        Args:
            verification_id: Verification ID
            
        Returns:
            Verification object
            
        Raises:
            VerinodeError: If verification not found or access denied
        """
        try:
            response = await self.http_client.get(f"/verifications/{verification_id}")
            
            if response.get("success"):
                verification_data = response["data"]
                verification = Verification(**verification_data)
                return verification
            else:
                error_msg = response.get("error", "Verification not found")
                raise VerinodeError(
                    error_msg,
                    error_code="VERIFICATION_NOT_FOUND"
                )
                
        except Exception as e:
            if isinstance(e, VerinodeError):
                raise
            raise VerinodeError(f"Failed to get verification {verification_id}: {str(e)}")
    
    async def update(
        self,
        verification_id: str,
        status: Optional[VerificationStatus] = None,
        comment: Optional[str] = None,
        evidence: Optional[Dict[str, Any]] = None
    ) -> Verification:
        """
        Update an existing verification.
        
        Args:
            verification_id: Verification ID
            status: New verification status
            comment: Optional comment
            evidence: Optional evidence
            
        Returns:
            Updated verification
            
        Raises:
            VerinodeError: If verification update fails
        """
        try:
            update_data = {}
            if status is not None:
                update_data["status"] = status.value
            if comment is not None:
                update_data["comment"] = comment
            if evidence is not None:
                update_data["evidence"] = evidence
            
            response = await self.http_client.patch(
                f"/verifications/{verification_id}",
                json=update_data
            )
            
            if response.get("success"):
                verification_data = response["data"]
                verification = Verification(**verification_data)
                self.logger.info(f"Updated verification: {verification.id}")
                return verification
            else:
                raise VerinodeError(
                    response.get("error", "Failed to update verification"),
                    error_code="VERIFICATION_UPDATE_FAILED"
                )
                
        except Exception as e:
            if isinstance(e, VerinodeError):
                raise
            raise VerinodeError(f"Failed to update verification {verification_id}: {str(e)}")
    
    async def delete(self, verification_id: str) -> bool:
        """
        Delete a verification.
        
        Args:
            verification_id: Verification ID
            
        Returns:
            True if deleted successfully
            
        Raises:
            VerinodeError: If verification deletion fails
        """
        try:
            response = await self.http_client.delete(f"/verifications/{verification_id}")
            
            if response.get("success"):
                self.logger.info(f"Deleted verification: {verification_id}")
                return True
            else:
                raise VerinodeError(
                    response.get("error", "Failed to delete verification"),
                    error_code="VERIFICATION_DELETE_FAILED"
                )
                
        except Exception as e:
            if isinstance(e, VerinodeError):
                raise
            raise VerinodeError(f"Failed to delete verification {verification_id}: {str(e)}")
    
    async def list(
        self,
        proof_id: Optional[str] = None,
        verifier_id: Optional[str] = None,
        status: Optional[VerificationStatus] = None,
        options: Optional[QueryOptions] = None
    ) -> PaginatedResponse:
        """
        List verifications with optional filtering and pagination.
        
        Args:
            proof_id: Filter by proof ID
            verifier_id: Filter by verifier ID
            status: Filter by verification status
            options: Query options (pagination, sorting, etc.)
            
        Returns:
            Paginated list of verifications
        """
        try:
            params = {}
            
            if proof_id:
                params["proof_id"] = proof_id
            if verifier_id:
                params["verifier_id"] = verifier_id
            if status:
                params["status"] = status.value
            
            if options:
                params.update({
                    "page": options.page,
                    "page_size": options.page_size,
                    "include_total": options.include_total
                })
                
                if options.filters:
                    for i, filter_obj in enumerate(options.filters):
                        params[f"filter_{i}_field"] = filter_obj.field
                        params[f"filter_{i}_operator"] = filter_obj.operator
                        params[f"filter_{i}_value"] = filter_obj.value
                
                if options.sort:
                    for field, direction in options.sort.items():
                        params[f"sort_{field}"] = direction
            
            response = await self.http_client.get("/verifications", params=params)
            
            if response.get("success"):
                data = response["data"]
                
                items = [Verification(**item) for item in data.get("items", [])]
                
                return PaginatedResponse(
                    items=items,
                    total_count=data.get("total_count", 0),
                    page=data.get("page", 1),
                    page_size=data.get("page_size", 10),
                    has_next=data.get("has_next", False),
                    has_prev=data.get("has_prev", False)
                )
            else:
                raise VerinodeError(
                    response.get("error", "Failed to list verifications"),
                    error_code="VERIFICATION_LIST_FAILED"
                )
                
        except Exception as e:
            if isinstance(e, VerinodeError):
                raise
            raise VerinodeError(f"Failed to list verifications: {str(e)}")
    
    async def approve(
        self,
        verification_id: str,
        comment: Optional[str] = None,
        evidence: Optional[Dict[str, Any]] = None
    ) -> Verification:
        """
        Approve a verification.
        
        Args:
            verification_id: Verification ID
            comment: Optional approval comment
            evidence: Optional approval evidence
            
        Returns:
            Updated verification
        """
        return await self.update(
            verification_id,
            status=VerificationStatus.APPROVED,
            comment=comment,
            evidence=evidence
        )
    
    async def reject(
        self,
        verification_id: str,
        comment: Optional[str] = None,
        evidence: Optional[Dict[str, Any]] = None
    ) -> Verification:
        """
        Reject a verification.
        
        Args:
            verification_id: Verification ID
            comment: Optional rejection comment
            evidence: Optional rejection evidence
            
        Returns:
            Updated verification
        """
        return await self.update(
            verification_id,
            status=VerificationStatus.REJECTED,
            comment=comment,
            evidence=evidence
        )
    
    async def get_statistics(
        self,
        proof_id: Optional[str] = None,
        verifier_id: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Get verification statistics.
        
        Args:
            proof_id: Optional proof ID filter
            verifier_id: Optional verifier ID filter
            
        Returns:
            Verification statistics
        """
        try:
            params = {}
            if proof_id:
                params["proof_id"] = proof_id
            if verifier_id:
                params["verifier_id"] = verifier_id
            
            response = await self.http_client.get("/verifications/statistics", params=params)
            
            if response.get("success"):
                return response["data"]
            else:
                raise VerinodeError(
                    response.get("error", "Failed to get verification statistics"),
                    error_code="VERIFICATION_STATS_FAILED"
                )
                
        except Exception as e:
            if isinstance(e, VerinodeError):
                raise
            raise VerinodeError(f"Failed to get verification statistics: {str(e)}")
    
    async def bulk_approve(
        self,
        verification_ids: List[str],
        comment: Optional[str] = None
    ) -> List[Verification]:
        """
        Approve multiple verifications at once.
        
        Args:
            verification_ids: List of verification IDs
            comment: Optional approval comment
            
        Returns:
            List of updated verifications
        """
        try:
            response = await self.http_client.post(
                "/verifications/bulk-approve",
                json={
                    "verification_ids": verification_ids,
                    "comment": comment
                }
            )
            
            if response.get("success"):
                verifications = [Verification(**v) for v in response["data"]]
                self.logger.info(f"Bulk approved {len(verifications)} verifications")
                return verifications
            else:
                raise VerinodeError(
                    response.get("error", "Failed to bulk approve verifications"),
                    error_code="VERIFICATION_BULK_APPROVE_FAILED"
                )
                
        except Exception as e:
            if isinstance(e, VerinodeError):
                raise
            raise VerinodeError(f"Failed to bulk approve verifications: {str(e)}")
    
    async def bulk_reject(
        self,
        verification_ids: List[str],
        comment: Optional[str] = None
    ) -> List[Verification]:
        """
        Reject multiple verifications at once.
        
        Args:
            verification_ids: List of verification IDs
            comment: Optional rejection comment
            
        Returns:
            List of updated verifications
        """
        try:
            response = await self.http_client.post(
                "/verifications/bulk-reject",
                json={
                    "verification_ids": verification_ids,
                    "comment": comment
                }
            )
            
            if response.get("success"):
                verifications = [Verification(**v) for v in response["data"]]
                self.logger.info(f"Bulk rejected {len(verifications)} verifications")
                return verifications
            else:
                raise VerinodeError(
                    response.get("error", "Failed to bulk reject verifications"),
                    error_code="VERIFICATION_BULK_REJECT_FAILED"
                )
                
        except Exception as e:
            if isinstance(e, VerinodeError):
                raise
            raise VerinodeError(f"Failed to bulk reject verifications: {str(e)}")

File: sdks/go/pkg/verinode/services/verification.go
package services

import (
	"context"
	"fmt"
	"log"

	"github.com/Great-2025/verinode-go/pkg/verinode/types"
)

// VerificationService handles verification-related operations
type VerificationService struct {
	httpClient HTTPClient
	config     ConfigProvider
}

// NewVerificationService creates a new verification service
func NewVerificationService(httpClient HTTPClient, config ConfigProvider) *VerificationService {
	return &VerificationService{
		httpClient: httpClient,
		config:     config,
	}
}

// Create creates a new verification
func (s *VerificationService) Create(ctx context.Context, req *types.VerificationCreateRequest) (*types.Verification, error) {
	var resp types.VerificationResponse
	err := s.httpClient.Post(ctx, "/verifications", req, &resp)
	if err != nil {
		return nil, fmt.Errorf("failed to create verification: %w", err)
	}
	
	if !resp.Success {
		return nil, fmt.Errorf("failed to create verification: %s", safeString(resp.Error))
	}
	
	if s.config.IsLoggingEnabled() {
		log.Printf("Created verification: %s", resp.Data.ID)
	}
	
	return &resp.Data, nil
}

// Get retrieves a verification by ID
func (s *VerificationService) Get(ctx context.Context, verificationID string) (*types.Verification, error) {
	var resp types.VerificationResponse
	err := s.httpClient.Get(ctx, fmt.Sprintf("/verifications/%s", verificationID), &resp)
	if err != nil {
		return nil, fmt.Errorf("failed to get verification %s: %w", verificationID, err)
	}
	
	if !resp.Success {
		return nil, fmt.Errorf("verification not found: %s", safeString(resp.Error))
	}
	
	return &resp.Data, nil
}

// Update updates an existing verification
func (s *VerificationService) Update(ctx context.Context, verificationID string, status *types.VerificationStatus, comment *string, evidence map[string]interface{}) (*types.Verification, error) {
	updateData := make(map[string]interface{})
	
	if status != nil {
		updateData["status"] = string(*status)
	}
	if comment != nil {
		updateData["comment"] = *comment
	}
	if evidence != nil {
		updateData["evidence"] = evidence
	}
	
	var resp types.VerificationResponse
	err := s.httpClient.Patch(ctx, fmt.Sprintf("/verifications/%s", verificationID), updateData, &resp)
	if err != nil {
		return nil, fmt.Errorf("failed to update verification %s: %w", verificationID, err)
	}
	
	if !resp.Success {
		return nil, fmt.Errorf("failed to update verification: %s", safeString(resp.Error))
	}
	
	if s.config.IsLoggingEnabled() {
		log.Printf("Updated verification: %s", resp.Data.ID)
	}
	
	return &resp.Data, nil
}

// Delete deletes a verification
func (s *VerificationService) Delete(ctx context.Context, verificationID string) error {
	var resp types.APIResponse
	err := s.httpClient.Delete(ctx, fmt.Sprintf("/verifications/%s", verificationID), &resp)
	if err != nil {
		return fmt.Errorf("failed to delete verification %s: %w", verificationID, err)
	}
	
	if !resp.Success {
		return fmt.Errorf("failed to delete verification: %s", safeString(resp.Error))
	}
	
	if s.config.IsLoggingEnabled() {
		log.Printf("Deleted verification: %s", verificationID)
	}
	
	return nil
}

// List lists verifications with optional filtering and pagination
func (s *VerificationService) List(ctx context.Context, proofID *string, verifierID *string, status *types.VerificationStatus, options *types.QueryOptions) (*types.PaginatedResponse, error) {
	params := make(map[string]string)
	
	if proofID != nil {
		params["proof_id"] = *proofID
	}
	if verifierID != nil {
		params["verifier_id"] = *verifierID
	}
	if status != nil {
		params["status"] = string(*status)
	}
	
	if options != nil {
		params["page"] = fmt.Sprintf("%d", options.Page)
		params["page_size"] = fmt.Sprintf("%d", options.PageSize)
		params["include_total"] = fmt.Sprintf("%t", options.IncludeTotal)
		
		for i, filter := range options.Filters {
			params[fmt.Sprintf("filter_%d_field", i)] = filter.Field
			params[fmt.Sprintf("filter_%d_operator", i)] = filter.Operator
			params[fmt.Sprintf("filter_%d_value", i)] = fmt.Sprintf("%v", filter.Value)
		}
		
		for field, direction := range options.Sort {
			params[fmt.Sprintf("sort_%s", field)] = fmt.Sprintf("%d", direction)
		}
	}
	
	var resp types.VerificationsResponse
	err := s.httpClient.Get(ctx, "/verifications", &resp, params...)
	if err != nil {
		return nil, fmt.Errorf("failed to list verifications: %w", err)
	}
	
	if !resp.Success {
		return nil, fmt.Errorf("failed to list verifications: %s", safeString(resp.Error))
	}
	
	return &resp.Data, nil
}

// Approve approves a verification
func (s *VerificationService) Approve(ctx context.Context, verificationID string, comment *string, evidence map[string]interface{}) (*types.Verification, error) {
	status := types.VerificationStatusApproved
	return s.Update(ctx, verificationID, &status, comment, evidence)
}

// Reject rejects a verification
func (s *VerificationService) Reject(ctx context.Context, verificationID string, comment *string, evidence map[string]interface{}) (*types.Verification, error) {
	status := types.VerificationStatusRejected
	return s.Update(ctx, verificationID, &status, comment, evidence)
}

// GetStatistics retrieves verification statistics
func (s *VerificationService) GetStatistics(ctx context.Context, proofID *string, verifierID *string) (map[string]interface{}, error) {
	params := make(map[string]string)
	
	if proofID != nil {
		params["proof_id"] = *proofID
	}
	if verifierID != nil {
		params["verifier_id"] = *verifierID
	}
	
	var resp types.StatisticsResponse
	err := s.httpClient.Get(ctx, "/verifications/statistics", &resp, params...)
	if err != nil {
		return nil, fmt.Errorf("failed to get verification statistics: %w", err)
	}
	
	if !resp.Success {
		return nil, fmt.Errorf("failed to get verification statistics: %s", safeString(resp.Error))
	}
	
	return resp.Data, nil
}

// BulkApprove approves multiple verifications at once
func (s *VerificationService) BulkApprove(ctx context.Context, verificationIDs []string, comment *string) ([]types.Verification, error) {
	req := map[string]interface{}{
		"verification_ids": verificationIDs,
	}
	if comment != nil {
		req["comment"] = *comment
	}
	
	var resp types.VerificationsResponse
	err := s.httpClient.Post(ctx, "/verifications/bulk-approve", req, &resp)
	if err != nil {
		return nil, fmt.Errorf("failed to bulk approve verifications: %w", err)
	}
	
	if !resp.Success {
		return nil, fmt.Errorf("failed to bulk approve verifications: %s", safeString(resp.Error))
	}
	
	// Convert interface{} slice to Verification slice
	verifications := make([]types.Verification, len(resp.Data.Items))
	for i, item := range resp.Data.Items {
		if verification, ok := item.(types.Verification); ok {
			verifications[i] = verification
		}
	}
	
	if s.config.IsLoggingEnabled() {
		log.Printf("Bulk approved %d verifications", len(verifications))
	}
	
	return verifications, nil
}

// BulkReject rejects multiple verifications at once
func (s *VerificationService) BulkReject(ctx context.Context, verificationIDs []string, comment *string) ([]types.Verification, error) {
	req := map[string]interface{}{
		"verification_ids": verificationIDs,
	}
	if comment != nil {
		req["comment"] = *comment
	}
	
	var resp types.VerificationsResponse
	err := s.httpClient.Post(ctx, "/verifications/bulk-reject", req, &resp)
	if err != nil {
		return nil, fmt.Errorf("failed to bulk reject verifications: %w", err)
	}
	
	if !resp.Success {
		return nil, fmt.Errorf("failed to bulk reject verifications: %s", safeString(resp.Error))
	}
	
	// Convert interface{} slice to Verification slice
	verifications := make([]types.Verification, len(resp.Data.Items))
	for i, item := range resp.Data.Items {
		if verification, ok := item.(types.Verification); ok {
			verifications[i] = verification
		}
	}
	
	if s.config.IsLoggingEnabled() {
		log.Printf("Bulk rejected %d verifications", len(verifications))
	}
	
	return verifications, nil
}
